File: Leetcode_with_Cpp/39combinationSum.cpp
class Solution {
private:
    vector<vector<int>> res;
    void backtrace(vector<int>& candidates, int target, vector<int>& p, int index) {
        if (target < 0) return;
        if (target == 0) {
            res.push_back(p);
            return;
        }
        for (int i = index; i < candidates.size(); i++) {
            p.push_back(candidates[i]);
            backtrace(candidates, target - candidates[i], p, i);
            p.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if (candidates.size() == 0) return res;
        vector<int> p;
        backtrace(candidates, target, p, 0);
        return res;
    }
};

File: Leetcode_with_py/39combinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if not candidates:
            return []
        if min(candidates) > target:
            return []
        candidates.sort()
        res = []
        def backtrace(candidates, target, temp_list):
            if target == 0:
                res.append(temp_list)
            if target < 0:
                return
            for i in range(len(candidates)):
                if candidates[i] > target:
                    break
                backtrace(candidates[i:], target - candidates[i], temp_list + [candidates[i]])
        backtrace(candidates, target, [])
        return res

# 回溯算法关键在于: 不合适就退回上一步
# 然后通过约束条件, 减少时间复杂度.